### PR TITLE
[WIP] [Guidance Wanted] Fix arm binaries compilation

### PR DESCRIPTION
**Pull request by jgabriel98**
## Bug description
The build toolchain and ppa packages tells it supports `armhf` which on debian seems to be `arm6` and `arm7` (_...don't ask me why_).
But when building for `armhf` (32bit arm), it was compiling `armv7l` binaries for `armv6l`, causing myst executable to crash with `"fatal error illegal instruction"`.

This same problem/bug happened on this other project: https://github.com/goreleaser/goreleaser/issues/2721
So my solution was lightly inspired on it.

## Results
Got it working on my _raspbery pizero W_, with both 
 - directly running `./myst` binaries (dynamic and static compilation)
 - generating `.deb` package and installing it

It was crosscompiled from a linux x86_64 machine.

------------------------

### Note
Currently this PR is incomplete and can only generate the
 - `armv6` binaries (see `buildBinaryFor()` and `packageStandalone()` changes)
 - `.deb` package (see `PackageLinuxDebianArmv6l()` changes)
 - static build binaries (see `PackageLinuxArmv6l()` changes)
 
It didn't know what more needed to be refactored, since i don´t know how what happens after artifacts gets uploaded to S3 bucket.

So any assistance is welcome.
Also feel free to reach on discord, i'm on the mysterium-network channel! (user: `#gaba3535`)

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -109,6 +109,11 @@ package:linux-debian-arm:
   tags: [go]
   script: go run mage.go -v PackageLinuxDebianArm
 
+package:linux-debian-armv6l:
```
Comment (jgabriel98) [inline .gitlab-ci.yml:112]: I actually didnt test this CI workflow.
The only reason i added this here was to follow the pattern.

**Comment by jgabriel98** [comment]
> @jgabriel98 hey, thanks you for contribution, we will be on this ASAP 🎉

You're welcome!
Let me know of any updates on this, if possible.

File: ci/packages/package.go [hunk 2]
```diff
@@ -341,21 +373,21 @@ func goGet(pkg string) error {
 	return sh.RunWith(map[string]string{"GO111MODULE": "off"}, "go", "get", "-u", pkg)
 }
 
-func packageStandalone(binaryPath, os, arch string) error {
+func packageStandalone(binaryPath, os, arch string, extraEnvs map[string]string) error {
```
Comment (mdomasevicius) [inline ci/packages/package.go:376]: Consider wrapping all of these arguments into a struct if it's not too much work

File: .gitlab-ci.yml (hunk 1 above)
Comment (mdomasevicius) [reply .gitlab-ci.yml:112]: Seems fine 👍

**Review by mdomasevicius — CHANGES_REQUESTED** [review]
Here is a test [pipeline](https://gitlab.com/mysteriumnetwork/node/-/pipelines/1134398142) for your changes

image

Your added step **2** works but it introduces a race: Steps **1** and **2** use same [script](https://github.com/mysteriumnetwork/node/blob/6a34402be5c89201d1f90204a452c2b8756a59f0/bin/package_debian#L46). Notice the package naming tempate:

> PACKAGE_FILE="${DIR_BUILD}/myst_linux_${ARCH}.deb"

As a result both steps produce artifact with same name:
> upload: build/package/myst_linux_armhf.deb to s3://mysteriumnetwork-node-1134398142/build-artifacts/myst_linux_armhf.deb

So which ever finishes last will be the one that should appear under [release assets](https://github.com/mysteriumnetwork/node/releases/tag/1.29.3-rc1)

@soffokl Please advise: I am not sure if changing `package_debian#49` to introduce more granular naming won't affect anything down the line: PPA release etc....

File: ci/packages/package.go (hunk 2 above)
Comment (jgabriel98) [reply ci/packages/package.go:376]: sure, no problems.
where do you suggest to declare the struct?

note: I'm a experience developer, but never touched golang before. So suggestions/tips related to golang patterns and good practices will always be wellcome

Comment (mdomasevicius) [reply ci/packages/package.go:376]: > when you say "wrapping all of these arguments" you mean (binaryPath, os, arch string, extraEnvs map[string]string)? or something else?
where do you suggest to declare the struct?

Meant exactly this ☝️ 🙂

File: ci/packages/package.go [hunk 3]
```diff
@@ -92,6 +104,26 @@ func PackageLinuxDebianArm() error {
 	return env.IfRelease(storage.UploadArtifacts)
 }
 
+// PackageLinuxDebianArm64 builds and stores debian armv6l package
```
Comment (soffokl) [inline ci/packages/package.go:107]:
```suggestion
// PackageLinuxDebianArm61 builds and stores debian armv6l package
```

File: ci/packages/package.go [hunk 4]
```diff
@@ -92,6 +104,26 @@ func PackageLinuxDebianArm() error {
...
+	}
+	envi := map[string]string{
+		"GOOS":   "linux",
+		"GOARCH": "arm",
+		"GOARM":  "6",
+	}
+	if err := sh.RunWith(envi, "bin/build"); err != nil {
+		return err
+	}
+	if err := packageDebian("build/myst/myst", "armhf"); err != nil {
```
Comment (soffokl) [inline ci/packages/package.go:121]: It looks like we just need to change the string here and we are ready to go.
```suggestion
	if err := packageDebian("build/myst/myst", "arm61"); err != nil {
```

**Comment by jgabriel98** [comment]
Hey there, sorry for taking that long.
I have a full time job as a developer, and didn't got to much free time to take a look into this.
I'll try to implement whats left on this week.

**Pull request closed without merging**